snuba_sdk/query_optimizers/or_optimizer.py

```diff
@@ -0,0 +1,68 @@
+from snuba_sdk.conditions import (
+    BooleanCondition,
+    BooleanOp,
+    Condition,
+    Op,
+)
+from snuba_sdk.formula import Formula
+from snuba_sdk.timeseries import Timeseries
+
+
+class OrOptimizer:
+    def optimize(self, query: Formula | Timeseries) -> Formula | Timeseries:
+        if query.filters is None:
+            return query
+
+        optimized = False
+        new_filters = []
+        for cond in query.filters:
+            res = self._optimize_condition(cond)
+            if res is not None:
+                optimized = True
+                new_filters.append(res)
+            else:
+                new_filters.append(cond)
+
+        if not optimized:
+            return query
+        elif isinstance(query, Timeseries):
+            return Timeseries(
+                metric=query.metric,
+                aggregate=query.aggregate,
+                aggregate_params=query.aggregate_params,
+                filters=new_filters,
+                groupby=query.groupby,
+            )
+        else:
+            return Formula(
+                function_name=query.function_name,
+                parameters=query.parameters,
+                aggregate_params=query.aggregate_params,
+                filters=new_filters,
+                groupby=query.groupby,
+            )
+
+    def _optimize_condition(
+        self, cond: BooleanCondition | Condition
+    ) -> BooleanCondition | Condition | None:
+        """
+        Given a condition, returns the optimized version, or None if it can't be optimized
+        """
+        if cond.op != BooleanOp.OR:
+            return None
+        shared_lhs = None
+        rhsides = []
+        for curr in cond.conditions:
+            if (
+                not isinstance(curr, Condition)
+                or curr.op != Op.EQ
+                or (shared_lhs and curr.lhs != shared_lhs)
+            ):
+                # can't be optimized
+                return None
+            if not shared_lhs:
+                shared_lhs = curr.lhs
+            rhsides += [curr.rhs]
+        assert shared_lhs
+        assert rhsides
+        return Condition(shared_lhs, Op.IN, rhsides)  # type: ignore
```

Comment:
Why is there a `# type: ignore` here?

Reply:
the new rhs for the `in` is a list of rhs, which works out to be of type
`Sequence[ScalarType | Expression]`
but rhs of a condition can only be of type
`Sequence[ScalarLiteralType | Expression]`

Reply:
There's something funky with what you're doing here then, there should be a way to have this type correctly.

Reply:
I don't think this is possible with the current type definitions. The best explanation I would be able to give for my reasoning over text would be what I wrote above.

RHS can be a sequence of literals, but it cannot be a sequence of sequence. Since a rhs can be a sequence, a list of rhs would be a sequence of sequence which is invalid.

Heres also a simple case to illustrate:
![Screenshot 2024-04-09 at 4 02 45 PM](https://github.com/getsentry/snuba-sdk/assets/24424170/2653f2e5-775c-4266-9163-193299b8586c)